crates/runtime/src/execution_plan/fallback_scan.rs

```diff
@@ -0,0 +1,217 @@
+use arrow::datatypes::SchemaRef;
+use async_trait::async_trait;
+use datafusion::datasource::TableProvider;
+use datafusion::error::{DataFusionError, Result};
+use datafusion::execution::context::SessionState;
+use datafusion::execution::{SendableRecordBatchStream, TaskContext};
+use datafusion::logical_expr::Expr;
+use datafusion::physical_plan::coalesce_partitions::CoalescePartitionsExec;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use datafusion::physical_plan::{
+    DisplayAs, DisplayFormatType, ExecutionPlan, ExecutionPlanProperties, Partitioning,
+    PlanProperties,
+};
+use futures::{stream, StreamExt};
+use std::any::Any;
+use std::fmt;
+use std::sync::Arc;
+
+#[derive(Clone)]
+#[allow(clippy::module_name_repetitions)]
+pub struct FallbackScanParams {
+    state: SessionState,
+    projection: Option<Vec<usize>>,
+    filters: Vec<Expr>,
+    limit: Option<usize>,
+}
+
+impl FallbackScanParams {
+    #[must_use]
+    pub fn new(
+        state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        filters: &[Expr],
+        limit: Option<usize>,
+    ) -> Self {
+        Self {
+            state: state.clone(),
+            projection: projection.cloned(),
+            filters: filters.to_vec(),
+            limit,
+        }
+    }
+}
+
+/// `FallbackScanExec` takes an input `ExecutionPlan` and a fallback `TableProvider`.
+/// If the input `ExecutionPlan` returns 0 rows, the fallback `TableProvider.scan()` is executed.
+///
+/// The input and fallback `ExecutionPlan` must have the same schema, execution modes and equivalence properties.
+#[allow(clippy::module_name_repetitions)]
+pub struct FallbackScanExec {
+    /// The input execution plan.
+    input: Arc<dyn ExecutionPlan>,
+    /// A closure to get the fallback execution plan if needed.
+    fallback_table_provider: Arc<dyn TableProvider>,
+    fallback_scan_params: FallbackScanParams,
+    properties: PlanProperties,
+}
+
+impl FallbackScanExec {
```

Comment:
Fallback is bit generic. It could mean the first priority of query encounters an error and fallback to the next one. Is this planning to handle all fallback cases?

Reply:
Renamed to `FallbackOnZeroResultsScanExec`